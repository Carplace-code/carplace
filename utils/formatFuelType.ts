import { FuelType } from "@prisma/client";

export default function formatFuelType(fuelType: FuelType): string {
  switch (fuelType) {
    case FuelType.gas:
      return "Gasolina";
    case FuelType.diesel:
      return "Diésel";
    case FuelType.electricity:
      return "Eléctrico";
    case FuelType.hybrid:
      return "Híbrido";
    default:
      return "Desconocido";
  }
}
