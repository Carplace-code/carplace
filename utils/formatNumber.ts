export default function formatNumber(value: number, options: { useThousandsSeparator?: boolean } = {}): string {
  if (typeof value !== "number") throw new TypeError("Value must be a number or Decimal instance");

  const { useThousandsSeparator } = options;
  const stringValue = value.toString();
  return useThousandsSeparator ? stringValue.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ".") : stringValue;
}
