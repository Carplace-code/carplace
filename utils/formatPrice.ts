export default function formatPrice(price: number | string): string {
  const localPrice = typeof price === "string" ? parseFloat(price.replace(/[^0-9.-]+/g, "")) : price;
  return Number(localPrice).toLocaleString("es-CL", {
    style: "currency",
    currency: "CLP",
  });
}
