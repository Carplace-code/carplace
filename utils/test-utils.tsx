// tests/test-utils.tsx
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import { ReactNode } from "react";

// Puedes extender esta función si necesitas envolver en más contextos
export function renderWithQueryClient(children: ReactNode) {
  const queryClient = new QueryClient();
  return render(<QueryClientProvider client={queryClient}>{children}</QueryClientProvider>);
}
