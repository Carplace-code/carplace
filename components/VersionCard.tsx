"use client";

import { useAuth } from "@clerk/nextjs";
import type { Prisma } from "@prisma/client";
import { Calendar, Fuel } from "lucide-react";
import Image from "next/image";
import Link from "next/link";

import { Button } from "@/components/ui/button";
import { Card, CardContent } from "@/components/ui/card";
import formatFuelType from "@/utils/formatFuelType";

import AddToWishlistModal from "./AddToWishlistModal";

export interface MinimumGridVersion
  extends Prisma.VersionGetPayload<{
    include: {
      model: {
        include: {
          brand: true;
        };
      };
      trims: {
        include: {
          carListings: {
            include: {
              images: true;
            };
          };
        };
      };
      wishlistItems: true;
    };
  }> {}

interface Props {
  version: MinimumGridVersion;
}

export default function VersionCard({ version: { id, model, trims, year } }: Props) {
  const { isSignedIn } = useAuth();
  const firstCarListingImageUrl = trims[0]?.carListings[0]?.images[0]?.url;
  const fuelTypes = [...new Set(trims.map((t) => formatFuelType(t.fuelType)))];
  const minPrice = Math.min(...trims.flatMap((t) => t.carListings.map((cl) => cl.price.toNumber())));
  const hasPrice = minPrice !== Number.POSITIVE_INFINITY;

  return (
    <Card className="group overflow-hidden transition-all duration-200 hover:shadow-lg">
      <CardContent className="p-0">
        {/* Image Container */}
        <div className="bg-muted relative aspect-video overflow-hidden">
          {firstCarListingImageUrl ? (
            <Image
              src={firstCarListingImageUrl || "/placeholder.svg"}
              alt={`${model.brand.name} ${model.name}`}
              fill
              className="object-cover transition-transform duration-200 group-hover:scale-105"
              sizes="(max-width: 640px) 100vw, (max-width: 1024px) 50vw, (max-width: 1280px) 33vw, 25vw"
            />
          ) : (
            <div className="text-muted-foreground flex h-full w-full items-center justify-center">
              <div className="text-center">
                <div className="bg-muted-foreground/10 mx-auto mb-2 flex h-12 w-12 items-center justify-center rounded-full">
                  <svg className="h-6 w-6" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M4 16l4.586-4.586a2 2 0 012.828 0L16 16m-2-2l1.586-1.586a2 2 0 012.828 0L20 14m-6-6h.01M6 20h12a2 2 0 002-2V6a2 2 0 00-2-2H6a2 2 0 00-2 2v12a2 2 0 002 2z"
                    />
                  </svg>
                </div>
                <p className="text-sm">Sin imagen</p>
              </div>
            </div>
          )}

          {/* Wishlist button overlay */}
          {isSignedIn && (
            <div className="absolute top-2 right-2">
              <AddToWishlistModal versionId={id} />
            </div>
          )}
        </div>

        {/* Content */}
        <div className="p-4">
          {/* Brand and Model */}
          <div className="mb-2">
            <p className="text-muted-foreground text-xs font-medium tracking-wide uppercase">{model.brand.name}</p>
            <h3 className="line-clamp-1 text-lg leading-tight font-semibold">{model.name}</h3>
          </div>

          {/* Year and Fuel Type */}
          <div className="text-muted-foreground mb-3 flex items-center gap-3 text-sm">
            <div className="flex items-center gap-1">
              <Calendar className="h-3 w-3" />
              <span>{year}</span>
            </div>
            {fuelTypes.length > 0 && (
              <div className="flex items-center gap-1">
                <Fuel className="h-3 w-3" />
                <span className="line-clamp-1">{fuelTypes.join(", ")}</span>
              </div>
            )}
          </div>

          {/* Price */}
          {hasPrice && (
            <div className="mb-4">
              <p className="text-muted-foreground text-sm">Desde</p>
              <p className="text-primary text-xl font-bold">${minPrice.toLocaleString()}</p>
            </div>
          )}

          {/* Action Button */}
          <Button asChild className="w-full" size="sm">
            <Link href={`/versions/${id}`}>Ver detalles</Link>
          </Button>
        </div>
      </CardContent>
    </Card>
  );
}
