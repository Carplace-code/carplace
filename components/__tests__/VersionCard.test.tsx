// components/__tests__/VersionCard.test.tsx
import { Decimal } from "@prisma/client/runtime/library";
import { describe, expect, it, vi } from "vitest";

import { renderWithQueryClient } from "@/utils/test-utils";

import type { MinimumGridVersion } from "../VersionCard";
import VersionCard from "../VersionCard";

// Mock de Clerk
vi.mock("@clerk/nextjs", () => ({
  useAuth: () => ({ isSignedIn: true }),
}));

describe("VersionCard", () => {
  it("renders correctly and matches snapshot", () => {
    const mockVersion: MinimumGridVersion = {
      id: "version-1",
      year: 2023,
      modelId: "model-1",
      model: {
        id: "model-1",
        name: "C200",
        bodyType: "sedan",
        brandId: "brand-1",
        brand: {
          id: "brand-1",
          name: "Mercedes-Benz",
        },
      },
      trims: [
        {
          id: "trim-1",
          name: "Base",
          motorSize: 2000,
          fuelType: "gas",
          transmissionType: "automatic",
          versionId: "version-1",
          carListings: [
            {
              id: "listing-1",
              sellerId: "seller-1",
              sourceId: "source-1",
              url: "https://example.com",
              title: "C200 2023",
              description: "En excelente estado",
              price: new Decimal(35990000),
              priceCurrency: "CLP",
              trimId: "trim-1",
              year: 2023,
              mileage: 10000,
              exteriorColor: "Silver",
              interiorColor: "Black",
              isNew: false,
              location: "Santiago",
              publishedAt: new Date(),
              scrapedAt: new Date(),
              images: [
                {
                  id: "img-1",
                  url: "https://example.com/image.jpg",
                  listingId: "listing-1",
                },
              ],
            },
          ],
        },
      ],
      wishlistItems: [],
    };

    const { container } = renderWithQueryClient(<VersionCard version={mockVersion} />);
    expect(container).toMatchSnapshot();
  });
});
