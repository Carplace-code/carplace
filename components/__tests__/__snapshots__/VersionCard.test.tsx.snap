// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VersionCard > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm group overflow-hidden transition-all duration-200 hover:shadow-lg"
    data-slot="card"
  >
    <div
      class="p-0"
      data-slot="card-content"
    >
      <div
        class="bg-muted relative aspect-video overflow-hidden"
      >
        <img
          alt="Mercedes-Benz C200"
          class="object-cover transition-transform duration-200 group-hover:scale-105"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="(max-width: 640px) 100vw, (max-width: 1024px) 50vw, (max-width: 1280px) 33vw, 25vw"
          src="/_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fexample.com%2Fimage.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
        <div
          class="absolute top-2 right-2"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 has-[>svg]:px-3 ml-2 h-auto p-1"
            data-slot="button"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-heart h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="p-4"
      >
        <div
          class="mb-2"
        >
          <p
            class="text-muted-foreground text-xs font-medium tracking-wide uppercase"
          >
            Mercedes-Benz
          </p>
          <h3
            class="line-clamp-1 text-lg leading-tight font-semibold"
          >
            C200
          </h3>
        </div>
        <div
          class="text-muted-foreground mb-3 flex items-center gap-3 text-sm"
        >
          <div
            class="flex items-center gap-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-calendar h-3 w-3"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 2v4"
              />
              <path
                d="M16 2v4"
              />
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="4"
              />
              <path
                d="M3 10h18"
              />
            </svg>
            <span>
              2023
            </span>
          </div>
          <div
            class="flex items-center gap-1"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-fuel h-3 w-3"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="3"
                x2="15"
                y1="22"
                y2="22"
              />
              <line
                x1="4"
                x2="14"
                y1="9"
                y2="9"
              />
              <path
                d="M14 22V4a2 2 0 0 0-2-2H6a2 2 0 0 0-2 2v18"
              />
              <path
                d="M14 13h2a2 2 0 0 1 2 2v2a2 2 0 0 0 2 2a2 2 0 0 0 2-2V9.83a2 2 0 0 0-.59-1.42L18 5"
              />
            </svg>
            <span
              class="line-clamp-1"
            >
              Gasolina
            </span>
          </div>
        </div>
        <div
          class="mb-4"
        >
          <p
            class="text-muted-foreground text-sm"
          >
            Desde
          </p>
          <p
            class="text-primary text-xl font-bold"
          >
            $
            35,990,000
          </p>
        </div>
        <a
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 w-full"
          data-slot="button"
          href="/versions/version-1"
        >
          Ver detalles
        </a>
      </div>
    </div>
  </div>
</div>
`;
